### PR TITLE
Implementing Ensemble Selection Techniques from PHEM Repository 

### PR DESCRIPTION
**Pull request by alee4977**
**Summary of Changes**

This pull request introduces changes to the AutoGluon framework that would support ensemble selection using methods from the PHEM repository in addition to the already implemented greedy ensemble selection algorithm. Key modifications include new classes that implement the logic for constructing an ensemble of models using CMA-ES, CMAES with Normalization, and Quality (Diversity) Optimization. 

**Detailed Changes**

CMAES Class (`CMAESSimulate`):
- Wrote a  `_fit()` function that uses the CMAES algorithm from the PHEM repository to take in model predictions and optimizes the weight vector for the ensemble model built using the `CMAES` class

    - Purpose: Use the CMA-ES algorithm to find the optimal weight vector for the ensemble fitted using prediction probabilities and ground truth labels
    - Functionality: The CMAESSimulate class is an extension of `EnsembleSelection`, because it uses a lot of the methods of `EnsembleSelection`, and the super class `AbstractWeightedEnsemble`. The `_fit()` method creates a list of fake fitted base models from the predictions array, where each row in the predictions array corresponds to a model and the columns correspond to samples. Creates a score metric using the appropriate evaluation metric based on the dataset the ensemble model is being built on.  The list of fake fitted models and score metric are passed as parameters to the `CMAES` class which is then used to fit the ensemble using the prediction probabilities and ground truth labels passed to the `fit()` method. Returns the weights for the fitted ensemble. 
    - Helpful Resource: https://github.com/LennartPurucker/phem/blob/master/examples/sklearn_examples/all_phem_methods_for_sklearn_models.py

CMAES with Normalization Class (`CMAESWithNormalizationSimulate`):
- Wrote a  `_fit()` function that uses the CMAES algorithm with normalization from the PHEM repository to take in model predictions and optimizes the weight vector for ensemble model built using the `CMAES` class

    - Purpose: Use the CMA-ES algorithm with normalization to find the optimal weight vector for the ensemble fitted using prediction probabilities and ground truth labels
    - Functionality: The `CMAESWithNormalizationSimulate` class is an extension of `EnsembleSelection`, because it uses a lot of the methods of `EnsembleSelection`, and the super class `AbstractWeightedEnsemble`. The _fit() method creates a list of fake fitted base models from the predictions array, where each row in the predictions array corresponds to a model and the columns correspond to samples. Creates a score metric using the appropriate evaluation metric based on the dataset ensemble model is being built on.  The list of fake fitted models and score metric are passed as parameters to the `CMAES` class which is then used to fit the ensemble using the prediction probabilities and ground truth labels passed to the `_fit()` method. Also the variable `normalize_weights` is set to "softmax" and the variable `trim_weights` is set to "ges-like". Returns the optimal weights for the fitted ensemble. 

Quality Optimization Class (`QualityOptimization`):
- Wrote a  `_fit()` function that uses the Quality Optimization algorithm from the PHEM repository to take in model predictions and optimizes the weight vector for ensemble model built using the `QDOEnsembleSelection` class

    - Purpose: Use the QO algorithm to find the optimal weight vector for the ensemble fitted using prediction probabilities and ground truth labels
    - Functionality: The `QualityOptimization` class is an extension of `EnsembleSelection`, because it uses a lot of the methods of `EnsembleSelection`, and the super class `AbstractWeightedEnsemble`. The `_fit()` method creates a list of fake fitted base models from the predictions array, where each row in the predictions array corresponds to a model and the columns correspond to samples. Creates a score metric using the appropriate evaluation metric based on the dataset ensemble model is being built on.  The list of fake fitted models and score metric are passed as parameters to the `QDOEnsembleSelection` class which is then used to fit the ensemble using the prediction probabilities and ground truth labels passed to the `_fit()` method. Also the `variable archive_type` is set to “quality”. Returns the optimal weights for the fitted ensemble. 

Quality Diversity Optimization Class (`QualityDiversityOptimization`):
- Wrote a  `_fit()` function that uses the Quality Diversity Optimization algorithm from the PHEM repository to take in model predictions and optimizes the weight vector for ensemble model built using the `QDOEnsembleSelection` class

     - Purpose: Use the QDO algorithm to find the optimal weight vector for the ensemble fitted using prediction probabilities and ground truth labels
     - Functionality: The `QualityDiversityOptimization` class is an extension of `EnsembleSelection`, because it uses a lot of the methods of `EnsembleSelection`, and the super class `AbstractWeightedEnsemble`. The `_fit()` method creates a list of fake fitted base models from the predictions array, where each row in the predictions array corresponds to a model and the columns correspond to samples. Creates a score metric using the appropriate evaluation metric based on the dataset ensemble model is being built on.  The list of fake fitted models and score metric are passed as parameters to the `QDOEnsembleSelection` class which is then used to fit the ensemble using the prediction probabilities and ground truth labels passed to the `_fit()` method. Also the variable behavior_space is set to the value returned from the `get_bs_ensemble_size_and_loss_correlation()`. Returns the optimal weights for the fitted ensemble. 

**Impact**

- Allows users to use the CMA-ES and Q(D)O methods to build ensembles of models in addition to the already implemented Greedy Ensemble Selection algorithm in AutoGluon.

**Results**

When any of the above classes are called, the corresponding ensemble method finds the best weight vector for the ensemble that is fitted using the prediction probabilities and ground truth values.

### CODE REVIEW

File: core/src/autogluon/core/models/greedy_ensemble/cmaes_normalization_simulate.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            cmaes_metric=msc(metric_name="rmse", is_binary=False, labels=list(range(2)))
+        elif (self.metric.name == "roc_auc"):
+            cmaes_metric=msc(metric_name="roc_auc", is_binary=True, labels=list(range(2)))
+        elif (self.metric.name == "log_loss"):
+            cmaes_metric=msc(metric_name="log_loss", is_binary=False, labels=list(range(2)))
+
+        # CMAES with Normalization 
+        cmaes_with_normalization = CMAES(
+            base_models=base_models_data,
+            n_iterations=10, # Number of iterations for ensemble selection
```
Comment (prateekdesai04) [inline core/src/autogluon/core/models/greedy_ensemble/cmaes_normalization_simulate.py:51]: why 10 and why not ensemble size ?

Comment (prateekdesai04) [inline core/src/autogluon/core/models/greedy_ensemble/cmaes_normalization_simulate.py:51]: why 10 and why not ensemble size ?

File: core/src/autogluon/core/models/greedy_ensemble/cmaes_normalization_simulate.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            score_metric=cmaes_metric,
+            random_state=1,
+            normalize_weights="softmax",
+            trim_weights="ges-like",
```
Comment (prateekdesai04) [inline core/src/autogluon/core/models/greedy_ensemble/cmaes_normalization_simulate.py:55]: `normalize` and `trim` weights is the only difference that I see in this file and `cmaes_simulate.py`
If so then why do we have redundant code?

File: core/src/autogluon/core/models/greedy_ensemble/qdo_ensemble_selection.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            qdo_metric=msc(metric_name="rmse", is_binary=False, labels=list(range(2)))
+        elif (self.metric.name == "roc_auc"):
+            qdo_metric=msc(metric_name="roc_auc", is_binary=True, labels=list(range(2)))
+        elif (self.metric.name == "log_loss"):
+            qdo_metric=msc(metric_name="log_loss", is_binary=False, labels=list(range(2)))
+
+        # QDO Method
+        qdo = QDOEnsembleSelection(
+            base_models=base_models_data,
+            n_iterations=10,
```
Comment (prateekdesai04) [inline core/src/autogluon/core/models/greedy_ensemble/qdo_ensemble_selection.py:54]: ditto

File: core/src/autogluon/core/models/greedy_ensemble/qo_ensemble_selection.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        # Determining the score metric 
+        if (self.metric.name == "root_mean_squared_error"):
+            qo_metric=msc(metric_name="rmse", is_binary=False, labels=list(range(2)))
+        elif (self.metric.name == "roc_auc"):
+            qo_metric=msc(metric_name="roc_auc", is_binary=True, labels=list(range(2)))
+        elif (self.metric.name == "log_loss"):
+            qo_metric=msc(metric_name="log_loss", is_binary=False, labels=list(range(2)))
+
+        # QO method
+        qo = QDOEnsembleSelection(
```
Comment (prateekdesai04) [inline core/src/autogluon/core/models/greedy_ensemble/qo_ensemble_selection.py:50]: there seems to be a lot of repetition of the same logic that is used in all 4 files apart from the minor differences when a CMAES or a QO object is created, any reason for this ?

**Comment by alee4977** [comment]
**Summary of Changes**

This pull request introduces changes to the AutoGluon framework that would support ensemble selection using CMAES, CMAES with Normalization, and Quality (Diversity) Optimization. Key modifications include a new class called EnsembleSelectionMethod. 

**Detailed Changes**

**1. Ensemble Selection Method Class (`EnsembleSelectionMethod`):**

- Wrote a `base_models_creation()` function: 

    - Purpose: Reduces the redundancy of the code, by having this function in one class instead of in four different simulation classes.  
    - Functionality: Creates a list of fake fitted base models from the predictions array, where each row in the predictions array corresponds to a model and the columns correspond to samples. 
- Wrote a `score_metric_creation()` function:
    - Purpose: Reduces the redundancy of the code, by having this function in one class instead of in four different simulation classes. 
    - Functionality: Returns an `AbtractMetric` object based on the score metric that is used for the dataset that the ensemble model is using.
- The `_fit()` method of the class creates the correct ensemble selection class according to the user’s input and returns the weights of the ensemble model. 

**Impact**

*  **Customizability:** Allows users to customize the ensemble selection method. 
*  **Modularization:** The code is modularized to reduce redundancy and increase code reusability. 

**Results**

The newly created EnsembleSelectionMethod class was able to replace four simulation classes that each corresponded to different ensemble selection methods.

**Comment by alee4977** [comment]
**Summary of Changes**
This pull request introduces changes to the `EnsembleSelectionMethod` class that reduces redundant code and makes the code more modularized. Key modifications include a file containing the functions that are called in the `_fit()` method of the class, and two new methods to create and validate ensembles using the CMAES and Q(D)O ensemble selection methods.

**Detailed Changes**

**1. Ensemble Selection Method Class (`EnsembleSelectionMethod`):**

* Wrote a `create_and_fit_ensemble_cmaes()` function:
    * Purpose: Prevents all the logic from being concentrated in one if-else block, and further modularizes the code.
    * Functionality: Creates and returns a fitted ensemble using the ensemble selection method (Single Best, CMAES, or CMAES with normalization) and validation data provided in the function arguments. 
* Wrote a `create_and_fit_ensemble_qdo()` function:
    * Purpose: Prevents all the logic from being concentrated in one if-else block, and further modularizes the code.
    * Functionality: Creates and returns a fitted ensemble using the ensemble selection method (Quality Optimization or Quality Diversity Optimization) and validation data provided in the function arguments.
* The `_fit()` method of the class calls different functions to create and fit ensembles using the user specified ensemble selection technique and subsequently returns the weights of the ensemble model. 


**Impact**

* **Customizability:** Allows users to customize the ensemble selection method used to build the ensemble model. 
* **Modularization:** The code is further modularized to reduce redundancy and increase code reusability.


**Results**
The logic for building and fitting the ensemble model in the EnsembleSelectionMethod class was initially concentrated in one if-else block in the `_fit()` function. This PR breaks up the logic into separation functions, that are called within the `_fit()` function.

**Comment by alee4977** [comment]
**Summary of Changes**
This pull request introduces changes to the phem_methods.py files that contains the methods used in the EnsembleSelectionMethod class. Key modifications are changes to the score_metric_creation() function that set requires_confidences=False for the log loss metric, because the prediction probabilities are being passed in directly to the ensemble selection methods from TabRepo. 

**Detailed Changes**

1. **phem_methods.py file:**

* Updated the score_metric_creation() function: 
    * Purpose: Updates the logic in order for the PHEM methods to work correctly
    * Functionality: Creates and returns an AbstractMetric object for binary log loss

**Results**
The logic for building a scoring object for log loss is changed, so that the code is compatible with the repository of PHEM that is downloaded using the pip install phem command, and the user does not have to make any changes to the PHEM repository.

**Comment by alee4977** [comment]
**Summary of Changes**
This pull request introduces changes to the `phem_methods.py` file. Key modifications include a function to specify the top N base models to be used for ensemble building and an update that allows the PHEM repository implementation of Ensemble Selection to be used.  

**Detailed Changes**

**1. phem_methods.py file:**

* Added `the get_top_N_predictions()` function:
    * Purpose: Allows the top N base models to be used instead of using all the base models, which can make the ensemble model more prone to overfitting.
    * Functionality: Takes in a parameter N, which specifies the number of top models from the model library that will be used to build the ensemble. 
* Modified the `create_and_fit_ensemble_greedy_ensemble_selection()` function:
    * Purpose: Allows PHEM’s implementation of Ensemble Selection to be used, which is useful for running experiments where AutoGluon and PHEM’s individual implementations of Ensemble Selection can be compared. This would show whether Ensemble Selection’s superior performance up until this point is maybe due to AutoGluon tricks such as rounding.
    * Functionality: Based on the ensemble selection method passed in, either AutoGluon or PHEM’s Ensemble Selection implementation is used.
